### PR TITLE
Refs #14091 -- Substituted parameters in connection.queries on SQLite.

### PR DESCRIPTION
**Pull request by aaugustin**
Refs #14091.

### CODE REVIEW

**Comment by aaugustin** [comment]
Alex and Jacob rejected that approach when I first submitted it on [ticket 14091](https://code.djangoproject.com/ticket/14091).

I still believe it would be a good thing for beginners who are learning with SQLite. We shouldn't inflict manual substitution of parameters in SQL on them just because they haven't figured out how to install a database server yet.

The implementation won't score points for elegance but I don't see anything wrong with it. The point of abstinence-only education about building SQL with string interpolation is to avoid substituting unescaped parameters in SQL. We're writing a framework to handle this sort of mess so our users don't have to. 

Here I'm asking the database to escape parameters. That's the safest approach I can think of. It doesn't suffer from the problems of PHP's `mysql_escape_string`, a reimplementation that doesn't handle all cases correctly.

File: django/db/backends/sqlite3/operations.py [hunk 1]
```diff
@@ -103,6 +103,41 @@ def drop_foreignkey_sql(self):
     def pk_default_value(self):
         return "NULL"
 
+    def _quote_params_for_last_executed_query(self, params):
+        """
+        Only for last_executed_query! Don't use this to execute SQL queries!
+        """
+        sql = 'SELECT ' + ', '.join(['QUOTE(?)'] * len(params))
```
Comment (timgraham) [inline django/db/backends/sqlite3/operations.py]: Do we have sufficient test coverage? In particular, I'm looking for a test which verifies this "QUOTE(?)" bit.

File: django/db/backends/sqlite3/operations.py [hunk 2]
```diff
@@ -103,6 +103,41 @@ def drop_foreignkey_sql(self):
...
+        # Native sqlite3 cursors cannot be used as context managers.
+        try:
+            return cursor.execute(sql, params).fetchone()
+        finally:
+            cursor.close()
+
+    def last_executed_query(self, cursor, sql, params):
+        # Python substitutes parameters in Modules/_sqlite/cursor.c with:
+        # pysqlite_statement_bind_parameters(self->statement, parameters, allow_8bit_chars);
+        # Unfortunately there is no way to reach self->statement from Python.
```
Comment (timgraham) [inline django/db/backends/sqlite3/operations.py]: from Python, so we quote and substitute parameters manually.

File: tests/test_runner/test_debug_sql.py [hunk 3]
```diff
@@ -61,28 +61,14 @@ def test_output_verbose(self):
...
-                '''FROM "test_runner_person" WHERE '''
-                '''"test_runner_person"."first_name" = %s' '''
-                '''- PARAMS = (u'*', u'error');'''),
-            ('''QUERY = u'SELECT COUNT(%s) AS "__count" '''
-                '''FROM "test_runner_person" WHERE '''
-                '''"test_runner_person"."first_name" = %s' '''
-                '''- PARAMS = (u'*', u'fail');'''),
-        ]
+    expected_outputs = [
+        '''SELECT COUNT('*') AS "__count" '''
```
Comment (timgraham) [inline tests/test_runner/test_debug_sql.py]: Meanwhile, I merged a conflicting patch. I think you just need to remove the quotes from around the star.

File: tests/test_runner/test_debug_sql.py [hunk 4]
```diff
@@ -61,28 +61,14 @@ def test_output_verbose(self):
...
+            '''FROM "test_runner_person" WHERE '''
```
Comment (timgraham) [inline tests/test_runner/test_debug_sql.py]: flake8 doesn't like the hanging indent here.

File: django/db/backends/sqlite3/operations.py [hunk 5]
```diff
@@ -103,6 +103,41 @@ def drop_foreignkey_sql(self):
...
+        if params:
+            if isinstance(params, (list, tuple)):
+                params = self._quote_params_for_last_executed_query(params)
+            else:
+                keys = params.keys()
+                values = tuple(params.values())
+                values = self._quote_params_for_last_executed_query(values)
+                params = dict(zip(keys, values))
+            return sql % params
+
```
Comment (timgraham) [inline django/db/backends/sqlite3/operations.py]: chop blank line?

File: django/db/backends/sqlite3/operations.py (hunk 1 above)
Comment (aaugustin) [inline django/db/backends/sqlite3/operations.py]: Right, I shall add one.

File: tests/test_runner/test_debug_sql.py (hunk 3 above)
Comment (aaugustin) [inline tests/test_runner/test_debug_sql.py]: Yes, I had seen the other PR, it had to go in first, thanks for taking care of that.

File: tests/test_runner/test_debug_sql.py (hunk 4 above)
Comment (aaugustin) [inline tests/test_runner/test_debug_sql.py]: i don't like flake8 :P

**Comment by aaugustin** [comment]
All fixed.

File: django/db/backends/sqlite3/operations.py (hunk 1 above)
Comment (aaugustin) [inline django/db/backends/sqlite3/operations.py]: See `test_parameter_quoting_on_sqlite` below.

File: tests/backends/tests.py [hunk 6]
```diff
@@ -388,8 +387,19 @@ def test_no_interpolation_on_sqlite(self):
         # This shouldn't raise an exception
         query = "SELECT strftime('%Y', 'now');"
         connection.cursor().execute(query)
-        self.assertEqual(connection.queries[-1]['sql'],
-            str_prefix("QUERY = %(_)s\"SELECT strftime('%%Y', 'now');\" - PARAMS = ()"))
+        self.assertEqual(connection.queries[-1]['sql'], query)
+
+    @unittest.skipUnless(connection.vendor == 'sqlite',
```
Comment (timgraham) [inline tests/backends/tests.py]: There is a separate test case `SQLiteTests` which contains SQLite specific tests - how do you feel about moving this there? In fact, I'd be in favor of moving all the backend specific tests to `tests/backends/test_<backend>.py`. Would you be +1 on that idea (as a separate issue)?

**Comment by timgraham** [comment]
Looks good (test failures are transient). I usually put "Refs #14091 -- `<message>`" on the first line of the commit message.

File: tests/backends/tests.py (hunk 6 above)
Comment (aaugustin) [inline tests/backends/tests.py]: `backends` contains a bunch of loosely related tests. They should at a minimum be split in distinct modules and perhaps to their own test apps or to other existing test apps.

**Pull request merged**